src/Transform/ONNX/ShapeInference.cpp

```diff
@@ -21,31 +22,88 @@ namespace onnx_mlir {
 
 namespace {
 
+bool hasDynamicOrUnknownShape(Type type) {
+  if (auto tensorType = dyn_cast<TensorType>(type))
+    return !tensorType.hasStaticShape();
+
+  if (type.isa<NoneType>())
+    return false;
+
+  if (auto seqType = dyn_cast<SeqType>(type))
+    return ShapedType::isDynamic(seqType.getLength()) ||
+           hasDynamicOrUnknownShape(seqType.getElementType());
+
+  if (auto optType = dyn_cast<OptType>(type))
+    return hasDynamicOrUnknownShape(optType.getElementType());
+
+  llvm_unreachable("unknown type");
+}
+
+// Returns failure if the op can be verified and failed verification.
+LogicalResult verifyOp(Operation *op) {
+  if (auto info = op->getName().getRegisteredInfo()) {
+    return info->verifyInvariants(op);
+  } else {
+    return success(); // Unregistered ops are unverifiable.
+  }
+}
+
+// Variant of RewriterBase::updateRootInPlace(op) which finalizes/cancels the
+// root update if the callback succeeds/fails.
+template <typename Callback = std::function<LogicalResult()>>
+LogicalResult tryUpdateRootInPlace(
+    Operation *op, PatternRewriter &rewriter, Callback &&callback) {
+  rewriter.startRootUpdate(op);
+  if (failed(callback())) {
+    rewriter.cancelRootUpdate(op);
+    return failure();
+  } else {
+    rewriter.finalizeRootUpdate(op);
+    return success();
+  }
+}
+
+// Returns failure if the op and its results are unchanged,
+// like RewritePattern::matchAndRewrite().
+// Calls shapeInfOp.emitOpError() if there is any actual failure.
+LogicalResult inferShapes(
+    ShapeInferenceOpInterface shapeInfOp, PatternRewriter &rewriter) {
+  return tryUpdateRootInPlace(shapeInfOp, rewriter, [&]() -> LogicalResult {
+    OperationFingerPrint before(shapeInfOp);
```

Comment:
You seem to be getting a lot of milage out of OperationFingerPrint. Its a handy trick